llvm/lib/Target/AIE/AIEMachineScheduler.cpp

```diff
@@ -479,6 +486,159 @@ int AIEPostRASchedStrategy::getMaxDeltaCycles(const SchedBoundary &Zone) const {
+      // the multi-slot and schedule the new instruction.
+      if (!checkSlotConflict(HR.getSelectedAltDescs().getOpcode(MI),
+                             NewMIAltOpcode, Formats))
+        continue;
+
+      // Release the multi-slot instruction from the scoreboard to check if
+      // any other alternate opcode in presence of the new instruction will
+      // not create a hazard.
+      HR.releaseFromScoreboard(*HR.getSelectedAltDescs().getDesc(MI),
+                               HR.getMemoryBanks(MI), MI->operands(),
+                               MI->getMF()->getRegInfo(),
+                               CurrCycle - ZoneSU->BotReadyCycle);
+
+      // Check if the new instuction can be scheduled after unscheduling
+      // the conflicting multi-slot instruction.
+      if (HR.getHazardType(NewMIAltDesc, HR.getMemoryBanks(NewMI),
+                           NewMI->operands(), NewMI->getMF()->getRegInfo(),
+                           DeltaCycle) !=
+          ScheduleHazardRecognizer::HazardType::NoHazard) {
+        // If the new instruction cannot be scheduled after unscheduling the
+        // mulit-slot revert back the state of scoreboard to original state
+        // and continue.
+
+        HR.emitInScoreboard(*HR.getSelectedAltDescs().getDesc(MI),
+                            HR.getMemoryBanks(MI), MI->operands(),
+                            MI->getMF()->getRegInfo(),
+                            CurrCycle - ZoneSU->BotReadyCycle);
+        continue;
+      }
+
+      // Emit the new instruction in the scoreboard. This will help us
+      // to check if the previously unscheduled multi-slot instruction
+      // can be scheduled in the same cycle, with an alternate opcode.
+      HR.emitInScoreboard(NewMIAltDesc, HR.getMemoryBanks(NewMI),
+                          NewMI->operands(), NewMI->getMF()->getRegInfo(),
+                          DeltaCycle);
+
+      // Check if the previously unscheduled multi-slot instruction
+      // can be rescheduled in presense of the new instruction in the
+      // same cycle, with a different opcode.
+      for (const auto AltOpcodeInside : *AltOpcodes) {
+        const MCInstrDesc &Desc = getTII(*Zone.DAG)->get(AltOpcodeInside);
+        if (HR.getHazardType(Desc, HR.getMemoryBanks(MI), MI->operands(),
+                             MI->getMF()->getRegInfo(), DeltaCycle) ==
+            ScheduleHazardRecognizer::HazardType::NoHazard) {
+          // Cache the information to mutate the instruction during bumpNode()
+          MutateInstruction.insert(
+              std::make_pair(NewMI, std::make_pair(MI, &Desc)));
+
+          // if the new instuction was a multi-slot instuction and it failed the
+          // general check for isAvailabeNode() this means we have not set the
+          // selected opcode for the instruction. Set the selected opcode for
+          // the instruction.
+          if (AlternateOpcodes->size() > 1)
```

Comment:
We have `AlternateOpcodes` and `AltOpcodes` that refer to different instructions. Maybe a more comprehensive name will help to follow the code.